### PR TITLE
core: add remote controller discovery with multicast

### PR DESCRIPTION
**Pull request by dpwiz**

### CODE REVIEW

File: src/Simplex/Chat/Controller.hs [hunk 1]
```diff
@@ -1061,6 +1061,8 @@ data RemoteCtrlError
   | RCEBadState -- ^ A session is in a wrong state for the current operation
   | RCEBusy -- ^ A session is already running
   | RCETimeout
+  | RCENoKnownControllers -- ^ No previously-contacted controllers to discover
+  | RCEControllerMismatch -- ^ Attempting to confirm a found controller with another ID
```
Comment (epoberezkin) [inline src/Simplex/Chat/Controller.hs:1065]: RCEBadState?

File: src/Simplex/Chat/Controller.hs [hunk 2]
```diff
@@ -1076,6 +1078,10 @@ data ArchiveError
 -- | Host (mobile) side of transport to process remote commands and forward notifications
 data RemoteCtrlSession
   = RCSessionStarting
+  | RCSessionSearching
```
Comment (epoberezkin) [inline src/Simplex/Chat/Controller.hs:1081]: RCSessionDiscovery?

File: src/Simplex/Chat/Controller.hs [hunk 3]
```diff
@@ -1101,6 +1107,7 @@ data RemoteCtrlSession
 
 data RemoteCtrlSessionState
   = RCSStarting
+  | RCSSearching
```
Comment (epoberezkin) [inline src/Simplex/Chat/Controller.hs:1110]: same

File: src/Simplex/Chat/Remote.hs [hunk 4]
```diff
@@ -342,19 +341,58 @@ liftRH rhId = liftError (ChatErrorRemoteHost (RHId rhId) . RHEProtocolError)
 
 -- * Mobile side
 
+-- ** Multicast
+
 findKnownRemoteCtrl :: ChatMonad m => m ()
-findKnownRemoteCtrl = undefined -- do
+findKnownRemoteCtrl = handleCtrlError "findKnownRemoteCtrl" $ do
+  knownCtrls <- withStore' getRemoteCtrls
+  pairings <- maybe (throwError $ ChatErrorRemoteCtrl RCENoKnownControllers) (pure . fmap (\RemoteCtrl {ctrlPairing} -> ctrlPairing)) $ nonEmpty knownCtrls
```
Comment (epoberezkin) [inline src/Simplex/Chat/Remote.hs:349]: some mess here a bit...

File: src/Simplex/Chat/Remote.hs [hunk 5]
```diff
@@ -342,19 +341,58 @@ liftRH rhId = liftError (ChatErrorRemoteHost (RHId rhId) . RHEProtocolError)
...
+  withRemoteCtrlSession_ $ maybe (Right ((), Just RCSessionStarting)) (\_ -> Left $ ChatErrorRemoteCtrl RCEBusy)
+  foundCtrl <- newEmptyTMVarIO
+  cmdOk <- newEmptyTMVarIO
+  action <- async $ handleCtrlError "findKnownRemoteCtrl.discover" $ do
+    atomically $ takeTMVar cmdOk
+    (RCCtrlPairing {ctrlFingerprint}, inv) <- timeoutThrow (ChatErrorRemoteCtrl RCETimeout) (networkIOTimeout * 4) . withAgent $ \a -> rcDiscoverCtrl a pairings
```
Comment (epoberezkin) [inline src/Simplex/Chat/Remote.hs:355]: let's just make another constant - no need to couple them

File: src/Simplex/Chat/Remote.hs [hunk 6]
```diff
@@ -342,19 +341,58 @@ liftRH rhId = liftError (ChatErrorRemoteHost (RHId rhId) . RHEProtocolError)
...
+    rc_ <- withStore' $ \db -> getRemoteCtrlByFingerprint db ctrlFingerprint
+    rc <- maybe (throwChatError $ CEInternalError "connecting with a stored ctrl") pure rc_
```
Comment (epoberezkin) [inline src/Simplex/Chat/Remote.hs:357]: this maybe (throw) is hard to read... I thin it's better just use \case in this case, without moving Maybe to ExceptT territory, they are different kind of error.

```suggestion
    withStore' (`getRemoteCtrlByFingerprint` ctrlFingerprint) >>= \case
      Nothing -> throwChatError $ CEInternalError "connecting with a stored ctrl"
      Just rc -> do
```

same as above

File: src/Simplex/Chat/Remote.hs [hunk 7]
```diff
@@ -554,11 +576,12 @@ verifyRemoteCtrlSession execChatCommand sessCode' = handleCtrlError "verifyRemot
...
 
-handleCtrlError :: ChatMonad m => Text -> m a -> m a
-handleCtrlError name action = action `catchChatError` \e -> do
-  logError $ name <> " remote ctrl error: " <> tshow e
-  cancelActiveRemoteCtrl True
-  throwError e
+handleCtrlError :: (HasCallStack, ChatMonad m) => Text -> m a -> m a
+handleCtrlError name action =
+  action `catchChatError` \e -> do
+    withFrozenCallStack $ logError $ name <> " remote ctrl error: " <> tshow e
```
Comment (epoberezkin) [inline src/Simplex/Chat/Remote.hs:582]: what's that

File: src/Simplex/Chat/Controller.hs (hunk 2 above)
Comment (dpwiz) [reply src/Simplex/Chat/Controller.hs:1081]: It was Discovery, but View displays it as "(searching)". Renamed to match other states in "-ing" form.

File: src/Simplex/Chat/Remote.hs (hunk 7 above)
Comment (dpwiz) [reply src/Simplex/Chat/Remote.hs:582]: fixing provenance in log messages. `handleCtrlError` is used multiple times, but source position tells the message comes from `handleCtrlError` itself which is useless.

this actually obsoletes the `name` argument.

**Pull request merged**